main/lsp/LSPTypechecker.cc

```diff
@@ -642,6 +641,55 @@ LSPTypechecker::SlowPathResult LSPTypechecker::runSlowPath(LSPFileUpdates &updat
     }
 }
 
+void LSPTypechecker::cacheOpenFiles(absl::Span<const ast::ParsedFile> indexed, std::vector<core::FileRef> openFiles,
+                                    WorkerPool &workers) {
+    if (openFiles.empty()) {
+        return;
+    }
+
+    auto &logger = *config->logger;
+    Timer timeit(logger, "slow_path.cache_open_files");
+
+    auto fileq = make_shared<ConcurrentBoundedQueue<int>>(openFiles.size());
+    for (int i = 0; i < openFiles.size(); i++) {
+        fileq->push(i, 1);
+    }
+
+    auto resultq = make_shared<BlockingBoundedQueue<std::vector<ast::ParsedFile>>>(std::max(workers.size(), 1));
+    workers.multiplexJob("cacheOpenFiles", [fileq, resultq, indexed]() {
+        vector<ast::ParsedFile> threadResult;
+        int job;
+
+        for (auto result = fileq->try_pop(job); !result.done(); result = fileq->try_pop(job)) {
+            if (result.gotItem()) {
+                // TODO(trevor): We traverse the whole indexed set for each open file. We could speed this up by
+                // pre-sorting `indexed` and restarting from the previous result, given that `openFiles` is also
+                // sorted. That way each worker would traverse `indexed` at most once.
+                auto it = absl::c_find_if(indexed, [job](auto &ast) { return ast.file.id() == job; });
```

Comment:
Rather than having `N=indexed.size()` threads all scanning the `indexed` vector individually `M=openFiles.size()` times, what about a single thread that scans `indexed` once, and checks to see whether `openFiles.contains(ast.file.id())`?

Would benefit from changing `openFiles` from a vector to a set.

Reply:
Yep, that's a good point. I had it in my mind that the best approach here would be keeping two pointers in each sorted vector and advancing them in a loop, but the set gives better performance without requiring a sort of the `indexed` vector.